examples/go/tagging/handler/handler.go

```diff
@@ -31,140 +32,145 @@ type vcConfig struct {
 }
 
 // The structure for a Cloud Event.
-type vmEvent struct {
-	Event event `json:"data"`
-}
-type event struct {
-	Vm vmEventArgument `json:"Vm"`
-}
-type vmEventArgument struct {
-	Vm managedObjectReference `json:"Vm"`
-}
-type managedObjectReference struct {
-	Type  string `json:"Type"`
-	Value string `json:"Value"`
+type incoming struct {
+	Data types.Event `json:"data,omitempty"`
 }
 
+var (
+	client *vsClient.Client
+	ctx    = context.Background()
+	cfg    vcConfig
+	sigCh  = make(chan os.Signal)
+)
+
 // Handle a function invocation
 func Handle(req handler.Request) (handler.Response, error) {
-	ctx := context.Background()
-
-	// Retrieve vCenter and tag info from the config file.
-	cfg, err := loadTomlCfg()
-	if err != nil {
-		log.Error(err)
-		return
+	// Connect to vSphere govmomi API if not already done. Also, load the required
+	// vcconfig.
+	if client == nil {
+		err := vsConnect()
+		if err != nil {
+			log.Fatalf("connect to vSphere failed: %v", err)
+			return handler.Response{
+				Body:       nil,
+				StatusCode: http.StatusBadRequest,
+			}, fmt.Errorf("connect to vSphere failed: %w", err)
+		}
 	}
 
-	// Incoming request object.
-	var event vmEvent
-	err := json.Unmarshal(req.Body, &event)
+	// Set up os signal handling to log out of vSphere.
+	signal.Notify(sigCh, syscall.SIGTERM, os.Interrupt)
```

Comment:
What is the purpose of signal handling here in `handler.go`? It seems that something else is handling os signals and shutting everything down, and we wouldn't need to do anything. Because you're right--I noticed that when a pod shut down, the authentication was also revoked and I didn't need to log out. 
```
2020/03/13 23:05:58 stderr: 2020/03/13 23:05:58 got signal: terminated,
2020/03/13 23:05:58 stderr: log out of vSphere
2020/03/13 23:05:58 SIGTERM received.. shutting down server in 10s
2020/03/13 23:05:58 Removing lock-file : /tmp/.lock
2020/03/13 23:05:58 stderr: 2020/03/13 23:05:58 [entrypoint] SIGTERM received.. shutting down server in 10s
2020/03/13 23:05:58 stderr: 2020/03/13 23:05:58 vSphere logout failed: govmomi api logout failed: ServerFaultCode: The session is not authenticated.
```

Should I just get rid of signal handling?

Reply:
Perhaps the template itself is handling it? It's not an issue to have multiple subscribers to signals. Weird err response from the logout...this could be due to the context being cancelled and govmomi taking care of it? Not sure though...

You could try to replicate in a minimalistic environment where you cancel the context and see what happens?

Reply:
Yeah, I can create a minimalist environment and see if I can replicate what is happening. 

Not sure if it helps, but I did find what you suggested to look for. I see that go-ttp is logging out in the Nimbus Events log: 
> 03/24/2020, 6:04:56 PM User VSPHERE.LOCAL\XXXXXXX@10.xxx.xx.xx logged out (login time: Wednesday, March 25, 2020 1:04:24 AM UTC, number of API invocations: 1, user agent: Go-http-client/1.1)

Will update my findings soon...

Reply:
In my code, every time `Handle()` function is called, I spin up a goroutine with `go signalHandle()`. Oops. 

The fix was to create another global variable, `once` of type `sync.Once`, and call `once.Do()` on `go signalHandle()`. That fixed the problem.